sycl/source/detail/config.hpp

```diff
@@ -186,10 +186,10 @@ static const std::array<std::pair<std::string, info::device_type>, 5>
 
 // Array is used by SYCL_DEVICE_FILTER and SYCL_DEVICE_ALLOWLIST
 static const std::array<std::pair<std::string, backend>, 5> SyclBeMap = {
-    {{"host", backend::host},
-     {"opencl", backend::opencl},
+    {{"opencl", backend::opencl},
      {"level_zero", backend::level_zero},
      {"cuda", backend::cuda},
+     {"host", backend::host},
```

Comment:
`host` can be used as both backend and device types.
Changing the order of matching string in backend is needed to avoid matching 'host' as the backend type incorrectly.
For example, `SYCL_DEVICE_FILTER=opencl:host`. The user specified the `host` device type in the `opencl` backend, but our filter will try to match host first and recognize this filter as the `host` backend.

Reply:
Not sure I understand the problem. Considering the example, shouldn't the first word `opencl` be matched against bakend names first? And only after that take a look at the second word `host` and try to match it against device types?
Am I missing something?

Reply:
> Not sure I understand the problem. Considering the example, shouldn't the first word `opencl` be matched against bakend names first? And only after that take a look at the second word `host` and try to match it against device types?
> Am I missing something?

Yes, that's what we want. But because `host` was the first entry in the backend list, and it matched `host' as the backend in the filter `opencl:host` incorrectly.
So changing the order, we can avoid this kind of problem.

Reply:
In the example `opencl:host`, `opencl` is a backend and `host` is device type. Could you please clarify how we can match device type against backend?

Reply:
For the given filter string `opencl:host`, we first try to match the backend.
Please note that backend, device_type, device_num are all optional, meaning we don't know if any entry is intended backend or device_type until we find the match.
For the previous matching order, `host` was listed before `opencl` in the possible backend list.
So, `host` is matched as a backend for the filter `opencl:host`.
The problem is that we can use `host` as for the backend or the device_type.
We can avoid this mismatching issue by reordering the matching order as I did in this PR.

Reply:
So, does it mean that a user can pass `backend` and `device type` in any order? e.g.
`SYCL_DEVICE_FILTER=opencl:cpu` and `SYCL_DEVICE_FILTER=cpu:opencl` mean the same thing? I had an impression that values are optional but cannot be reordered.

It seems that for `SYCL_DEVICE_FILTER=cpu:opencl` case current parsing implementation just ignores `cpu` device type and sets to any device type. 

If they cannot be reordered I would suggest splitting filter to N ( N <= 3) values and try to match them one by one with backend, device type and id.
E.g.:

1. "opencl:cpu" -> "opencl", "cpu"
2. Match first value against list of BEs -> Matched!
3. Match the second value against list of device types -> Matched!
4. Id value is empty -> Any ID

1. "cpu:1" -> "cpu", "1"
2. Match first value against list of BEs -> No match
3. Match the first value against device types -> Matched!
4. The second value must be device ID(if we have an order) -> Convert to a number.

Reply:
> SYCL_DEVICE_FILTER=opencl:cpu and SYCL_DEVICE_FILTER=cpu:opencl mean the same thing?

No. 
As we defined in the documentation and error messages, we expect users to pass in the order of
1. backend
2. device_type
3. device_id

But, users can set it "cpu:opencl" by mistake, and we won't accept it as "opencl:cpu".

> If they cannot be reordered I would suggest splitting filter to N ( N <= 3) values and try to match them one by one with backend, device type and id.

Yes, we already match the filter entries in the order you mentioned. (i.e., backend, device_type, device_id)
The problem is that all three entries are **optional.**
So, we can't assume the first entry is always the backend.
For example, "gpu" or "cpu:1" or "2" are valid filter specification.
In these cases, we try to match with given backend list {opencl, level_zero, host, *}.
None of them matches, so we consider the backend type is "*".

But, "opencl:host" is the headache because "host" can be backend or device_type.
In the previous order of backend list {host, opencl, level_zero, *}, we always try to match the filter with `host` backend, and it is matched.
By reordering the backend to match with "opencl" first rather than "host", we can avoid this mismatch issue.

Reply:
I think I understand what you mean now.
Instead of parsing the entire filter string for the backend first, split the filter into N substrings.
And try to match with backend, device_type, device_num.
I will change as you suggested.

Reply:
Done. Thanks.